edbterraform/data/terraform/aws/modules/specification/variables.tf

```diff
@@ -62,7 +62,8 @@ variable "spec" {
       type          = optional(string)
       image_name    = string
       count         = optional(number, 1)
-      spot_max_price = optional(number)
+      spot          = optional(bool)
+      spot_instance_interruption_behavior = optional(string)
```

Comment:
Since this is mostly for cost-savings and short-runs should we remove `spot_instance_interruption_behavior = optional(string)` and keep the default as `stop` within the machine module? If we want to keep the override, we will need to do a validation check to avoid using `terminate` when using additional volumes.

Reply:
I'm not sure...  IIRC, you originally found the default in conflict with some specific instance types, so I think we need the override.

I've already forgotten, did we run into issues on the planning or apply phase?  If it fails on planning, do you still think we need to add our own check?

Reply:
These are the conflicts found:
* `one-time` request type causes ebs deletion to fail
* `terminate` not supported with request type `persistent`
* `hibernate` not supported with instance type `c5d.18xlarge`

`stop` with `persistent` worked for both `c5d.18xlarge` and `c5.4xlarge` with attached ebs volumes.

---

We ran into the issues during the `apply` phase so we will need to add our own check. 
If we would like to continue with validations, it can be a variable validation or a resource lifecycle precondition validation.
I think variable validations are limiting since it can only reference itself and we cannot use a local or another variable within its conditionals.

Here is an example within `resource "aws_instance" "machine"` with a precondition:
```terraform
  lifecycle {
    precondition {
      condition     = (
        var.machine.spec.spot == null || var.machine.spec.spot == false ||
        (var.machine.spec.spot == true && var.machine.spec.spot_instance_interruption_behavior != "terminate")
      )
      error_message = "Invalid configuration option"
    }
    # AMI is ignored because the data source
    # forces the resource to be re-created when apply is used again
    ignore_changes = [ami]
  }
```

Error message:
```bash
user@su22:~/Projects/edb-terraform/terraform
$ terraform_1.3.6 plan
╷
│ Error: Resource precondition failed
│ 
│   on modules/machine/main.tf line 59, in resource "aws_instance" "machine":
│   59:       condition     = (
│   60:         var.machine.spec.spot == null || var.machine.spec.spot == false ||
│   61:         (var.machine.spec.spot == true && var.machine.spec.spot_instance_interruption_behavior != "terminate")
│   62:       )
│     ├────────────────
│     │ var.machine.spec.spot is true
│     │ var.machine.spec.spot_instance_interruption_behavior is "terminate"
│ 
│ Invalid configuration option

```